src/main/java/org/terasology/math/geom/Polygon.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2014 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.terasology.math.geom;
+
+import java.util.List;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+
+/**
+ * @author Martin Steiger
+ */
+public final class Polygon implements Shape {
+
+    private final List<ImmutableVector2f> vertices;
+
+    /**
+     * @param vertices a list of vertices (copied for internal storage)
+     */
+    public Polygon(List<? extends BaseVector2f> vertices) {
+        Builder<ImmutableVector2f> bldr = ImmutableList.builder();
+        for (BaseVector2f v : vertices) {
+            if (v instanceof ImmutableVector2f) {
```

Comment:
I would suggest giving ImmutableVectors a static creation method (and set the constructor to private) that takes an existing vector, and have the logic to reuse/actually create the vector in that.